src/Nethermind/Nethermind.Trie/Pruning/TrieStoreDirtyNodesCache.cs

```diff
@@ -225,127 +225,151 @@ public void PruneCache(
+                    // we keep it to prevent key removal from removing it from DB.
+                    if (node.LastSeen == -1 || forceRemovePersistedNodes)
                     {
-                        // If its persisted and has last seen meaning it was recommitted,
-                        // we keep it to prevent key removal from removing it from DB.
-                        if (node.LastSeen == -1 || forceRemovePersistedNodes)
-                        {
-                            if (_logger.IsTrace) _logger.Trace($"Removing persisted {node} from memory.");
-
-                            Hash256? keccak = node.Keccak;
-                            if (keccak is null)
-                            {
-                                TreePath path2 = key.Path;
-                                keccak = node.GenerateKey(_trieStore.GetTrieStore(key.Address), ref path2, isRoot: true);
-                                if (keccak != key.Keccak)
-                                {
-                                    throw new InvalidOperationException($"Persisted {node} {key} != {keccak}");
-                                }
-
-                                node.Keccak = keccak;
-                            }
-                            Remove(key);
-
-                            Metrics.PrunedPersistedNodesCount++;
-                            continue;
-                        }
-
-                        if (_trieStore.IsNoLongerNeeded(node))
-                        {
-                            if (_logger.IsTrace) _logger.Trace($"Removing {node} from memory (no longer referenced).");
-                            if (node.Keccak is null)
-                            {
-                                throw new InvalidOperationException($"Removed {node}");
-                            }
+                        if (_logger.IsTrace) LogPersistedNodeRemoval(node);
 
-                            Metrics.PrunedPersistedNodesCount++;
-
-                            Remove(key);
-                            continue;
-                        }
+                        Hash256? keccak = (node.Keccak ??= GenerateKeccak(key, node));
+                        RemoveNodeFromCache(key, node);
+                        continue;
                     }
-                }
-                else if (_trieStore.IsNoLongerNeeded(node))
-                {
-                    if (_logger.IsTrace) _logger.Trace($"Removing {node} from memory (no longer referenced).");
-                    if (node.Keccak is null)
+
+                    if (_trieStore.IsNoLongerNeeded(node))
                     {
-                        throw new InvalidOperationException($"Removed {node}");
+                        RemoveNodeFromCache(key, node);
+                        continue;
                     }
+                }
+            }
+            else if (_trieStore.IsNoLongerNeeded(node))
+            {
+                RemoveNodeFromCache(key, node);
```

Comment:
Metric sould be different between the two `RemoveNodeFromCache`

Reply:
https://github.com/NethermindEth/nethermind/pull/8456